### PR TITLE
Compatibility v13

### LINKED ISSUE #54: [FEATURE] TYPO3 13 support
Just as info, #53 provides TYPO3 v13 compatibility and testing functionality in the backend is welcome.
Any hints for improvements are naturally welcome too.

### PR DESCRIPTION
**Pull request by DavidBruchmann**
fix #54

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -22,9 +22,10 @@
 		"issues": "https://github.com/kephson/paste-reference/issues"
 	},
 	"require": {
-		"php": "^8.1",
-		"typo3/cms-backend": "^12.4",
-		"typo3/cms-core": "^12.4"
+		"ext-pdo": "*",
+		"php": "^8.2",
+		"typo3/cms-backend": "^13.4 || dev.main",
```
Comment (PKuhlmay) [inline composer.json:27]: `dev.main` is wrong here.
It should be `dev-main` instead.

Personally I would leave it with `^13.4` only as this versions comes out in a week. Or change it to `^13.3 || dev-main` (_only during development, I would remove the `dev-main` constraint for the production release_) as with `^13.4 || dev-main` it isn't installable with the current sprint release 13.3 right?

BTW the **dev.main** leads to a failing pipeline at the moment.

File: ext_emconf.php [hunk 2]
```diff
@@ -11,10 +11,10 @@
     'title' => 'Paste reference for content elements',
     'description' => 'Paste reference instead of copy for content elements',
     'category' => 'plugin',
-    'version' => '3.0.0',
+    'version' => '4.0.0',
     'constraints' => [
         'depends' => [
-            'typo3' => '12.4.0-12.4.99',
+            'typo3' => '13.4.0-13.5.99',
```
Comment (PKuhlmay) [inline ext_emconf.php:17]: According to the roadmap https://typo3.org/cms/roadmap the 13.4 will be the LTS release.

```suggestion
            'typo3' => '13.4.0-13.4.99',
```

File: Classes/DataHandler/AbstractDataHandler.php [hunk 3]
```diff
@@ -26,7 +26,7 @@
 use Doctrine\DBAL\Driver\Exception as DBALDriverException;
 use Doctrine\DBAL\Exception as DBALException;
 use EHAERER\PasteReference\Helper\Helper;
-use TYPO3\CMS\Core\Database\Connection;
+use TYPO3\CMS\Core\Database\Connection as Typo3Connection;
```
Comment (PKuhlmay) [inline Classes/DataHandler/AbstractDataHandler.php:29]: Is there a reason to set is as `Typo3Connection`? There is no other `Connection` inside this class.

File: Documentation/DeveloperManual/Version-Comparison.ods
Comment (PKuhlmay) [inline Documentation/DeveloperManual/Version-Comparison.ods:1]: What is in this file? I can assume that this is a list of the versions, but I cannot open this file on my machine, and the file is also not referenced within the documentation.

**Review by PKuhlmay — CHANGES_REQUESTED** [review]

File: Documentation/DeveloperManual/Version-Comparison.ods
Comment (DavidBruchmann) [reply Documentation/DeveloperManual/Version-Comparison.ods:1]: at's an open office spreadsheet file, like excel

File: composer.json
Comment (PKuhlmay) [inline composer.json:1]: There are a lot of wrong versions under require-dev. For example the TYPO3 versions should be the same as under "require" and not ^12.4

File: Classes/DataHandler/AbstractDataHandler.php (hunk 3 above)
Comment (DavidBruchmann) [reply Classes/DataHandler/AbstractDataHandler.php:29]: mhm, artists inspiration could serve as only argument

File: composer.json (hunk 1 above)
Comment (DavidBruchmann) [reply composer.json:27]: > ^13.4 || dev-main it isn't installable

mhm, not sure. Just `^13.3 || dev-main` should be fine for the moment, another adjustment in 1-2 weeks is no problem.

File: ext_emconf.php (hunk 2 above)
Comment (DavidBruchmann) [reply ext_emconf.php:17]: good, that you clarified the versions!

File: Documentation/DeveloperManual/Version-Comparison.ods
Comment (DavidBruchmann) [reply Documentation/DeveloperManual/Version-Comparison.ods:1]: I'm open what to do with it, for the moment I mark this as resolved. Could be a separated issue.

File: composer.json
Comment (DavidBruchmann) [reply composer.json:1]: thanks, is fixed, at least in the typo3 namespace

File: composer.json [hunk 4]
```diff
@@ -22,42 +22,43 @@
...
+		"typo3/cms-lowlevel": "^13.3",
+		"typo3/cms-opendocs": "^13.3",
+		"typo3/cms-recycler": "^13.3",
+		"typo3/cms-redirects": "^13.3",
+		"typo3/cms-reports": "^13.3",
+		"typo3/cms-rte-ckeditor": "^13.3",
+		"typo3/cms-scheduler": "^13.3",
+		"typo3/cms-seo": "^13.3",
+		"typo3/cms-setup": "^13.3",
 		"typo3/cms-styleguide": "^12.0.5",
```
Comment (PKuhlmay) [inline composer.json:58]: this should also be another version: 
typo3/cms-styleguide: ^13.3
```suggestion
		"typo3/cms-styleguide": "^13.3"
```

Comment (DavidBruchmann) [reply composer.json]: thanks!! fixed.

Comment (DavidBruchmann) [reply composer.json]: thanks, is fixed

**Comment by DavidBruchmann** [comment]
@simonschaufi yes, I just tried to push a commit for it, but would need additional access rights for it.

**Comment by DavidBruchmann** [comment]
Concerning the failing test, I don't understand the issue.
The complaint is about this line:
```
if (count($jsLines) > 0) {
```
and saying this:

> ------ -------------------------------------------------------------------- 
>   Line   Classes/Hooks/PageLayoutController.php                              
>  ------ -------------------------------------------------------------------- 
>   98     Comparison operation ">" between int<1, 4> and 0 is always true.    
>          💡 Because the type is coming from a PHPDoc, you can turn off this  
>             check by setting treatPhpDocTypesAsCertain: false in your        
>             Build/phpstan/phpstan.neon.                                      
>  ------ --------------------------------------------------------------------

**Comment by simonschaufi** [comment]
Please don't merge main back into this branch but use rebase instead.

**Comment by DavidBruchmann** [comment]
I'll do a complete new PR based on the latest merge, so this here won't be used in this form.

**Pull request closed without merging**